---
title: "R Notebook"
output: html_notebook
---

```{r}
# from applyTallies help
library(h5vc)
library(BSgenome.Hsapiens.UCSC.hg19)

files <- c("NRAS.AML.bam","NRAS.Control.bam")
bamFiles <- file.path( system.file("extdata", package = "h5vcData"), files)
chrom = "1"
startpos <- 115247090
endpos <- 115259515
theData <- applyTallies( bamFiles, 
                         reference = Hsapiens[["chr1"]][startpos:endpos], 
                         chr = chrom, start = startpos, stop = endpos, 
                         ncycles = 10 )
str(theData)
```


## h5vc â€“ Scalabale nucleotide tallies using HDF5
Source: http://www.bioconductor.org/packages/devel/bioc//vignettes/h5vc/inst/doc/h5vc.tour.html

In this document we will illustrate the use of the h5vc package for creating and analysing nucleotide tallies of next generation sequencing experiments.

`h5vc` is a tool that is designed to provide researchers with a more intuitive and effective way of interacting with data from large cohorts of samples that have been sequenced with next generation sequencing technologies.

This file format will store nucleotide tallies rather than alignments and allow for easy and efficient real-time random access to the data of a whole cohort of samples. The details are described in the following section.

### Nucleotide Tally Definition

The tally data structure proposed here consists of 5 datasets that are stored for each chromosome (or contig). Those datasets are:

* Counts: A table that contains the number of observed mismatches at any combination of base, sample, strand and genomic position,
* Coverages: A table that contains the number of reads overlapping at any combination of sample, strand and genomic position
* Deletions: A Table that contains the number of observed deletions of bases at any combination of sample, strand and genomic position
* Insertions: A Table that contains the number of observed insertions of bases at any combination of sample, strand and genomic position (showing insertions following the position)
* Reference: A Table that contains the reference base against which the calls in the 'Deletions' and 'Counts' table were made.

We outline the basic layout of this set of tables here:

    Name	Dimensions	Datatype
    Counts	[ #bases, #samples, #strands, #positions ]	int
    Coverages	[ #samples, #strands, #positions ]	int
    Deletions	[ #samples, #strands, #positions ]	int
    Insertions	[ #samples, #strands, #positions ]	int
    Reference	[ #positions ]	int
    
An HDF5 file has an internal structure that is similar to a file system, where groups are the directories and datasets are the files. The layout of the tally file is as follows:

![](./images/hdf5_tally.png)

A tally file can contain data from more than one study but each study will reside in a separte tree with a group named with the study-ID at its root and sub-groups for all the chromosomes / contigs that are present in the study. Attached to each of the chromosome groups are the 4 datasets described above.

Additionally each chromsome group stores sample data about the samples involved in the experiment (patientID, type, location in the sample dimension) as HDF5 attributes. Convenience functions for extracting the metadata are provided, see examples below.

### A practical example


The first thing we do is set up the session by loading the h5vc and rhdf5 packages and finding the location of the example tally file.
```{r}
library(h5vc)
library(rhdf5)

tallyFile <- system.file( "extdata", "example.tally.hfs5", package = "h5vcData" )
```


We can inspect the data contained in this file with the h5ls function provided by rhdf5.

```{r}
h5ls(tallyFile)
```

In the resulting output we can find the different groups and datasets present in the file and we can extract the relevant sample data attached to those groups in the following way.

```{r}
sampleData <- getSampleData( tallyFile, "/ExampleStudy/16" )
sampleData
```

The sampleData object is a data.frame that contains information about the samples whose nucleotide tallies are present in the file. We can modify this object (e.g. add new columns) and write it back to the file using the setSampleData function, but we must be aware that a certain set of columns have to be present (Sample, Patient, Column and Type).


```{r}
# add a column to /ExampleStudy/16
sampleData$ClinicalVariable <- rnorm(nrow(sampleData))
setSampleData( tallyFile, "/ExampleStudy/16", sampleData )
sampleData
```

Now that we can find the sample metadata in the file it is time to extract some of the nuclotide tally data stored there. We can use two functions to achieve this, `h5readBlock` can be used to read a specified block of data along a given dimension (e.g. a region along the genomic position) and `h5dapply` can be used to apply a function in a blockwise fashion along a specified dimension (e.g. calculating coverage in bins of a certain size along the genomic position dimension).

We can read out a block of data in the following way:

```{r}
data <- h5readBlock(
  filename = tallyFile,
  group = "/ExampleStudy/16",
  names = c( "Coverages", "Counts" ),
  range = c(29000000, 29001000)
  )
str(data)
```
