---
title: "S4 classes"
output: html_notebook
---

FUll example of using S4 to create a gene sequence class.

Let's start by loading the package and read a fasta sequence 

```{r}
read.csv("./inst/extdata/aDnaSeq.fasta", header = FALSE)
```

```{r}
read.csv("./inst/extdata/moreDnaSeqs.fasta", header = FALSE)
```



## Define the super class and sub classes
Source: https://github.com/cran/sequences/blob/master/R/DataClasses.R

```{r}
setClass("GenericSeq",
         representation = representation(
           id = "character",
           alphabet = "character",
           sequence =  "character",
           "VIRTUAL"),
         validity = function(object) {
           isValid <- TRUE
           if (length(object@sequence)>0) {
             chars <- casefold(unique(unlist(strsplit(object@sequence,""))))
             isValid <- all(chars %in% casefold(object@alphabet))
           }
           if (!isValid)
             cat("Some characters are not defined in the alphabet.\n")
           return(isValid)
         })


setClass("DnaSeq",
         contains="GenericSeq",
         prototype = prototype(
           id = paste("my DNA sequence",date()),
           alphabet = c("A","C","G","T"),
           sequence = character())
         )

setClass("RnaSeq",
         contains="GenericSeq",
         prototype = prototype(
           id = paste("my RNA sequence",date()),
           alphabet = c("A","C","G","U"),
           sequence = character())
         )

# constructors for DNA and RNA
DnaSeq <- function(id, sequence) 
    new("DnaSeq", id = id, sequence = sequence)

RnaSeq <- function(id, sequence) 
    new("RnaSeq", id = id, sequence = sequence)
```


## Define the generics
Source: https://github.com/cran/sequences/blob/master/R/AllGenerics.R

```{r}
setGeneric("id", function(object, ...) standardGeneric("id"))
setGeneric("id<-", function(object,value) standardGeneric("id<-"))
setGeneric("alphabet", function(object, ...) standardGeneric("alphabet"))

## There is already a 'seq' method (see ?seq),
## although not a generic one (see isGeneric(seq))
setGeneric("seq", function(...) standardGeneric("seq"))
setGeneric("seq<-", function(object,value) standardGeneric("seq<-"))
## Same note that above for print
setGeneric("print", function(x, ...) standardGeneric("print"))

setGeneric("rev",function(x) standardGeneric("rev"))
setGeneric("comp",function(object, ...) standardGeneric("comp"))

setGeneric("transcribe", function(object, ...) standardGeneric("transcribe"))
```

## Methods for DNA and RNA
Source: https://github.com/cran/sequences/blob/master/R/methods-DnaSeq.R

```{r}
setMethod("comp", "DnaSeq",
          function(object, ...) {
            chartr("ACGT","TGCA",seq(object))
          })


setMethod("transcribe", "DnaSeq",
          function(object, ...) {
            .sequence <- chartr("T","U",toupper(seq(object)))
            .id <-  paste(id(object),"-- transcribed")
            rna <- new("RnaSeq",
                       id=.id,
                       alphabet=c("A","C","G","U"),
                       sequence=.sequence)
            return(rna)
          })

setMethod("comp", "RnaSeq",
          function(object, ...) {
            chartr("ACGU","UGCA",seq(object))
          })
```

## Generic sequence methods
Source: https://github.com/cran/sequences/blob/master/R/methods-GenSeq.R

```{r}
setMethod("show",
          "GenericSeq",
          function(object) {
            cat("Object of class",class(object),"\n")
            cat(" Id:",id(object),"\n")
            cat(" Length:",length(object),"\n")
            cat(" Alphabet:",alphabet(object),"\n")
            cat(" Sequence:",seq(object), "\n")
          })


setMethod("print",
          "GenericSeq",
          function(x) {
            sq <- strsplit(seq(x), "")[[1]]
            cat(">",id(x),"\n")
            cat(" 1   ")
            for (i in 1:length(x)) {
              if ((i %% 10)==0) {
                cat("\n")
                cat(i,"  ")
              }
              cat(sq[i])
            }
            cat("\n")
          })


setMethod("id", "GenericSeq", function(object, ...) object@id)
setMethod("id<-", "GenericSeq",
          function(object,value) object@id <- value)

# setReplaceMethod("id",
#                  signature(object="GenericSeq",
#                            value="character"),
#                  function(object, value) {
#                    object@id <- value
#                    if (validObject(object))
#                      return(object)
#                  })


setMethod("alphabet", "GenericSeq", function(object, ...) object@alphabet)
setMethod("length", "GenericSeq", function(x) nchar(x@sequence))

setMethod("seq", "GenericSeq", function(object, ...) object@sequence)

setReplaceMethod("seq",
                 signature(object="GenericSeq",
                           value="character"),
                 function(object, value) {
                   object@sequence <- value
                   if (validObject(object))
                     return(object)
                 })

setMethod("rev","GenericSeq",
          function(x) paste(rev(strsplit(seq(x),"")[[1]]),collapse=""))
          

## this is only an example of initialize function,
## note Generic is in fact virtual          
setMethod("initialize", "GenericSeq",
          function(.Object, ..., id="", sequence=""){
            .Object@id <- id
            .Object@sequence <- toupper(sequence)
            callNextMethod(.Object, ...)
          })
	

setMethod("[","GenericSeq",
          function(x, i, j="missing",drop="missing") {
            if (any(i > length(x)))
              stop("subscript out of bounds")
            s <- seq(x)
            s <- paste(strsplit(s,"")[[1]][i], collapse="")
            x@sequence <- s
            if (validObject(x))
              return(x)
          })
```

## Function to read sequences

```{r}
readFasta <- function(infile){
  lines <- readLines(infile)
  header <- grep("^>", lines)
  cat("Sequences found at lines: ", header, "\n")
  if (length(header)>1) {
    warning("Reading first sequence only.\n")
    lines <- lines[header[1]:(header[2]-1)]
    header <- header[1]
  }
  .id <- sub("^> *","",lines[header],perl=TRUE)
  .sequence <- toupper(paste(lines[(header+1):length(lines)],collapse=""))
  .alphabet <- toupper(unique(strsplit(.sequence,"")[[1]]))
  if (all(.alphabet %in% c("A","C","G","T"))) {
    # newseq <- new("DnaSeq",
    #               id=.id,
    #               sequence=.sequence)
    newseq <- DnaSeq(.id, .sequence)
    
  } else if (all(.alphabet %in% c("A","C","G","U"))) {
    # newseq <- new("RnaSeq",
    #               id=.id,
    #               sequence=.sequence)
      newSeq <- RnaSeq(.id, .sequence)
  } else {
    stop("Alphabet ",.alphabet," is unknown.")
  }
  if (validObject(newseq))
    return(newseq)
}
```

## Read 1st sequence

```{r}
fastafilename <- dir(path="./inst/extdata",
                     full.name=TRUE,
                     pattern="fasta$")
fastafilename
myseq <- readFasta(fastafilename[1])
myseq
```

```{r}
transcribe(myseq)
```

```{r}
alphabet(myseq)
```


```{r}
seq(myseq)
```

```{r}
print(myseq)
```

```{r}
rev(myseq)
```

```{r}
comp(myseq)
```

```{r}
length(myseq)
```

```{r}
myseq[5:10]
```

```{r}
myseq[5, 10]
# T
```

```{r}
myseq[1, 10]
# A
```

## Read the 2nd seqence

```{r}
myseq2 <- readFasta(fastafilename[2])
myseq2
```


```{r}
# set the id using setReplaceMethod
id(myseq2) <- "Sequence #1"
myseq2
```

```{r}
# get the id
id(myseq2)
```


```{r}
# using setReplaceMethod
`id<-`(myseq2, "this is sequence no. 1")
```


